### DISCUSSION TITLE
Combine Fieldwork in RL vs CP

### DISCUSSION THREAD

**Discussion by rohne83** [Courseplay/Courseplay_FS22#590] [Show and tell]
With the following lines, I'll try to explain how combines are working the fields in RL. At least that's what is practiced in eastern Germany. But doesn't seem like the approach is differing significantly in other parts of the world. But feel free to add your knowledge, thoughts or differing observations. Maybe some of the points not covered by CP yet might find the way into the mod anytime in the future.
I'll also add another comment with some pictures, as visualizing often helps to understand better ;)

### Headlands
In general headlands are created **around the whole field**, but not always. Especially if the field/rows are worked in normal up/down mode it can happen that at least one long edge is left out.
**Number** of headland lanes depends on working width and turning radius and of course the shape of the field. Three lanes is a good number in practice. Sometimes 4 might be needed, while with 12-15m headers even two lanes can be enough.
In most cases the **first headland pass is worked counterclockwise** - with the pipe over the fruit. The bottleneck ist the first combine anyway and this one cannot unload on the move on the first headland. But there are several other points to think of, most of which in favour of working counterclockwise: i.e. the first lane is the most critical one with obstacles at the edge or even illegal garbage, and you don't want to risk your header drive (situated on the left side of the header) to be damaged there. And as you have all monitors, switches and monitors in the right part of the cockpit, it's easier to also have the field edge to the right, so everything needed is in your view without having to turn the head. Older combines built before the 90s also had fixed pipes which would get stuck on every obstacle if driving clockwise.
If you have a convoy, the other combines would just follow in the same direction.
After first headland pass is finished, and additional headland lanes are required, the combine(s) will turn 180 degrees and work the **rest of the headland clockwise**. The reasons for working counterclockwise on the first lane are much less important there, but it's now possible to unload on the move.
CP is not able to work first and following headland in different direction, but you can still generate a course with headland passes counterclockwise and still have the fieldrows worked in correct direction. When you have a convoy of three or more combines, one headland pass is enough anyway. In FS 19 you just had to take care about the correct lane offset (which would vary between counterclock-headland and spirals/lands-mode field rows). Even in counterclock headlands CP combines make room for unloading without the unloader having to drive into the crop. Just the pocket in the first counterclock lane is not created yet (opposed to clockwise headlands where CP combines indeed create a pocket) if I remember correctly.

### Field rows
The pattern for the field rows depends on several factors. In general three patterns are common:
1. normal **up/down** - only used for smaller fields if there's just one combine and that one is _self-unloading_ into trailers parked at the field edge. You often try to place the trailers in the headland of a short edge, as you want to keep ways short and the combines prefer integrating the unloading into the field row turn.
2. **lands** - on larger fields where you have a _convoy_ of combines and/or _active unloaders_, up/down is not an option at all. Instead you split the field into different lands, and start each land by working from the inside out - just as CP lands is doing already. One significant difference to CP: CP is continuing this inside-out approach until the last lane. But the already harvested breach gets wider and wider with every pass. Once this breach is wider than the smaller of the two remaining fruit blocks, in RL this smaller block will be finished from outside in (as CP's spirals mode would do for the whole field). Mainly to reduce turning distances. Once finished, the next land will be started from inside out again.
3. **spirals** - exactly what CP spirals mode is doing. Driving distances during row turns and for the unloaders are a factor to be thought about of course. But if the field is very narrow it makes sense to just work the whole field outside-in, same way as the small blocks in lands mode are finished.

For up/down and sometimes spirals it's not always necessary to have exactly straight rows. The drivers would follow the _contours_ of the long field edges instead.
For starting new lands the fertilizer's tramlines are used as orientation. Or you have GPS which will always guide you into the exact position.
In any case combine drivers try to use any of their headland lanes during turning, as soil compaction is a big issue especially with modern big machinery. CP Pathfinder already has this feature as well (just limited to use only the outernmost headland lane only).

**Comment by rohne83** [comment]
**Some pictures for the headlands:**

![20210812_145227](https://user-images.githubusercontent.com/71030138/150003051-865a34df-b2bc-496a-b96e-8109a784d52f.jpg)
Starting on the first headland counter-clockwise with pipe directed towards field center.

![20210812_150119](https://user-images.githubusercontent.com/71030138/150003151-fdfdb50e-747f-4690-bc56-a1b8295b250f.jpg)
Unloading in the headland. The combine just needed to drive a bit backwards and to the side (this was the third cominbe in the convoy, thus the headland is quite large already).

![fsScreen_2022_01_18_15_25_23](https://user-images.githubusercontent.com/71030138/150003253-c79b2c9b-1923-4181-b396-01e5794ab4c7.png)
This unloading also works with CP already. Only the pocket for the first combine might still be an issue.

[This video](https://www.youtube.com/watch?v=y1qHq_vXELE) shows the relevant situations from the air. At 3:00 the combine creates its pocket for unloading, at 9:00 it reaches the end of the first headland, performs a U turn, and continues the remaining headland lanes clockwise.

**And some pictures for the fieldrows:**

![20200801_210805](https://user-images.githubusercontent.com/71030138/150004188-c95fd740-b501-4621-9d29-2f0bc6338c46.jpg)
Quite small and especially narrow rectangular field (just about 150m x 600m, the area behind is a completely different field). Would be just one land anyway. The combines work it completely in spirals mode.

![20210821_143928](https://user-images.githubusercontent.com/71030138/150004378-78c2d240-96db-46c4-9b12-763897ce1c53.jpg)
Self unloading combine running normal up/down rows. And the trailers parked in a good location so that unloading can be integrated into the headland turn. Looking closer you can also notice the field rows not being straight but following the contour of the field edge.

![20210812_160359](https://user-images.githubusercontent.com/71030138/150004758-d7313043-c44e-4635-86c6-5cb0730745b3.jpg)
On larger fields with active unloaders and/or convoys of combines where spirals is not an option, lands is the way to go. Tramlines used for orientation and the rest is done by GPS or Auto pilot.

![20210812_162422](https://user-images.githubusercontent.com/71030138/150004949-51eab814-a111-41c0-baea-b5959a6434e2.jpg)
After just 2 up and down passes for each combine, the harvested breach got a width of more than 100m already. About as wide as the remaining fruit block to the right, which was then finished in spirals.

![20210812_181123](https://user-images.githubusercontent.com/71030138/150005157-1d98d35a-c76b-44f3-b202-3384bd7942be.jpg)
Other combines on another field, but same pattern. To the left you see part of the larger fruit block, then the harvested breach. The combines are finishing the narrow block to the right in spirals mode currently, before they start a new land.

**Reply by Lactic68** [reply]
Thank you so much @rohne83 for your great and illustrated explanations of all different type of rows and headlands.
This help a lot to understand which methodology to use in different scenarios.

By the way, could you explain what are the differences between these CP headland corner modes please ?

Headland corner - smooth - will generate a simple smooth corner
Headland corner - round - will generate a round corner, the radius depends on the set turn radius, set on the value setting page on the HUD

As far as I can tell, both are using curves to manage headland corners.

**Comment by rohne83** [comment]
Basically I would always use sharp corners, that's the only way to work a field completely - be it real life or with CP.
The smooth and round corner options are rather for guys who want to save time, or if you have asymmetric towed implements such as plows where reversing  (which is unavoidable in sharp corners) might not be the best idea. It's your own choice then, where you want to have the left out spots: with round corners the innermost headland lane using the minimum turn radius, and this turning radius increases for every further lane towards the field edge (basic geometry). That way there aren't spots left out between the lanes, but there can be a quite huge area left out beyond the outermost headland lane.
Whereas for smooth corners, the minimum turn radius is applied to every headland lane. There are less areas left out, but you always have spots between the lanes.
@pvaiko had shown a nice graphic depicting this in any issue for FS19. Will try to find that one.

My personal approach is that one: for cultivating, seeding, fertilizing etc on mission fields I use smooth corners. Doesn't look too nice, but saves time and once the mission is completed, the whole field automatically gets converted to 100% worked by the Giants engine. For harvesting I normally use sharp corners, and also for nearly every other field work on my own fields (I don't use plows at all). Round corners instead are my choice for fields owned by me, when I think sharp is too time consuming or when I experienced still left out spots even with sharp. But I then use round for the whole cycle, starting especially with the seeding. That way, I will waste a bit of the field area which will not be sown, but I always get perfect work results and don't have issues with missed spots between the lanes.

**Comment by pvaiko** [comment]
> Whereas for smooth corners, the minimum turn radius is applied to every headland lane. 

Actually, there's no turn radius applied, just a generic smoothing algorithm.

**Comment by Ulanawius** [comment]
Thank you very much for the great work. :-)

I just realized, that you implemented the function, that the combine searches for an unloading vehicle while turning on the headland, when nearly full. This is something I was whishing for a long time.
Thank you again!
![proShot_2023-08-28_17-30-21](https://github.com/Courseplay/Courseplay_FS22/assets/76711921/411e2f1a-2d33-4cda-81b6-990db8bdb574)
